File: 0002/addTwoNumbers.php
<?php

use runtime\ListNode;

class Solution {

    /**
     * @param ListNode $l1
     * @param ListNode $l2
     * @return ListNode
     */
    function addTwoNumbers($l1, $l2) {
        $cur = new ListNode(0);
        $carry = 0;

        while ($l1 || $l2 || $carry) {
            $x = $l1 ? $l1->val : 0;
            $y = $l2 ? $l2->val : 0;
            $sum = $x + $y + $carry;
            $carry = floor($sum / 10);

//            echo "x = $x, y = $y, sum = $sum, carry = $carry\n";

            $node = new ListNode($sum % 10);

//            echo "val = $node->val\n";

            $cur->next = $node;
            $cur = $cur->next;

            $l1 = $l1 ? $l1->next : $l1;
            $l2 = $l2 ? $l2->next : $l2;
        }

        return $cur->next;
    }
}

File: 0002/addTwoNumbers.go
package _0002

import "github.com/raw34/leetcode/runtime"

func addTwoNumbers(l1 *runtime.ListNode, l2 *runtime.ListNode) *runtime.ListNode {
    head := &runtime.ListNode{}
    curr := head
    carry := 0
    for l1 != nil || l2 != nil {
        val1, val2 := 0, 0
        if l1 != nil {
            val1 = l1.Val
            l1 = l1.Next
        }
        if l2 != nil {
            val2 = l2.Val
            l2 = l2.Next
        }
        sum := val1 + val2 + carry
        carry = sum / 10
        sum = sum % 10
        curr.Next = &runtime.ListNode{Val: sum}
        curr = curr.Next
    }

    if carry > 0 {
        curr.Next = &runtime.ListNode{Val: carry}
    }

    return head.Next
}
